File: src/main/java/kr/devport/api/domain/wiki/repository/WikiSectionChunkRepositoryImpl.java
package kr.devport.api.domain.wiki.repository;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import kr.devport.api.domain.wiki.entity.WikiSectionChunk;
import lombok.RequiredArgsConstructor;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.jdbc.core.RowMapper;
import org.springframework.jdbc.core.ConnectionCallback;
import org.springframework.stereotype.Repository;

import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.List;
import java.util.Map;

@Repository
@RequiredArgsConstructor
public class WikiSectionChunkRepositoryImpl implements WikiSectionChunkRepositoryCustom {

    private static final int HNSW_EF_SEARCH = 64;

    private static final String SELECT_CHUNK_COLUMNS = """
            SELECT
                c.id,
                c.project_external_id,
                c.section_id,
                c.subsection_id,
                c.chunk_type,
                c.content,
                c.token_count,
                c.metadata,
                c.commit_sha
            """;

    private static final String VECTOR_BY_PROJECT_SQL = SELECT_CHUNK_COLUMNS + """
            , 1 - (c.embedding <=> cast(? AS vector)) AS score
            FROM wiki_section_chunks c
            WHERE c.project_external_id = ?
            ORDER BY c.embedding <=> cast(? AS vector)
            LIMIT ?
            """;

    private static final String VECTOR_GLOBAL_SQL = SELECT_CHUNK_COLUMNS + """
            , 1 - (c.embedding <=> cast(? AS vector)) AS score
            FROM wiki_section_chunks c
            ORDER BY c.embedding <=> cast(? AS vector)
            LIMIT ?
            """;

    private static final String LEXICAL_SQL = SELECT_CHUNK_COLUMNS + """
            ,
            (similarity(c.content, ?) + 0.5 * COALESCE(similarity(c.metadata->>'titleKo', ?), 0)) AS score
            FROM wiki_section_chunks c
            WHERE c.project_external_id = ?
              AND (c.content % ? OR c.metadata->>'titleKo' % ?)
            ORDER BY score DESC
            LIMIT ?
            """;

    private final JdbcTemplate jdbcTemplate;
    private final ObjectMapper objectMapper = new ObjectMapper();

    private final RowMapper<ScoredChunkRow> scoredChunkRowMapper = (rs, rowNum) ->
            new ScoredChunkRow(mapChunk(rs), rs.getDouble("score"));

    @Override
    public List<ScoredChunkRow> findSimilarChunksWithScore(String projectExternalId, String queryEmbedding, int limit) {
        return jdbcTemplate.execute((ConnectionCallback<List<ScoredChunkRow>>) connection -> {
            setVectorSearchEf(connection);
            try (PreparedStatement statement = connection.prepareStatement(VECTOR_BY_PROJECT_SQL)) {
                statement.setString(1, queryEmbedding);
                statement.setString(2, projectExternalId);
                statement.setString(3, queryEmbedding);
                statement.setInt(4, limit);
                try (ResultSet rs = statement.executeQuery()) {
                    return mapRows(rs);
                }
            } finally {
                resetVectorSearchEf(connection);
            }
        });
    }

    @Override
    public List<ScoredChunkRow> findSimilarChunksGlobalWithScore(String queryEmbedding, int limit) {
        return jdbcTemplate.execute((ConnectionCallback<List<ScoredChunkRow>>) connection -> {
            setVectorSearchEf(connection);
            try (PreparedStatement statement = connection.prepareStatement(VECTOR_GLOBAL_SQL)) {
                statement.setString(1, queryEmbedding);
                statement.setString(2, queryEmbedding);
                statement.setInt(3, limit);
                try (ResultSet rs = statement.executeQuery()) {
                    return mapRows(rs);
                }
            } finally {
                resetVectorSearchEf(connection);
            }
        });
    }

    @Override
    public List<ScoredChunkRow> findLexicalCandidates(String projectExternalId, String question, int limit) {
        return jdbcTemplate.query(
                LEXICAL_SQL,
                scoredChunkRowMapper,
                question,
                question,
                projectExternalId,
                question,
                question,
                limit
        );
    }

    private List<ScoredChunkRow> mapRows(ResultSet rs) throws SQLException {
        List<ScoredChunkRow> rows = new java.util.ArrayList<>();
        int rowNum = 0;
        while (rs.next()) {
            rows.add(scoredChunkRowMapper.mapRow(rs, rowNum++));
        }
        return rows;
    }

    private WikiSectionChunk mapChunk(ResultSet rs) throws SQLException {
        return WikiSectionChunk.builder()
                .id(rs.getLong("id"))
                .projectExternalId(rs.getString("project_external_id"))
                .sectionId(rs.getString("section_id"))
                .subsectionId(rs.getString("subsection_id"))
                .chunkType(rs.getString("chunk_type"))
                .content(rs.getString("content"))
                .tokenCount(rs.getInt("token_count"))
                .metadata(readMetadata(rs.getObject("metadata")))
                .commitSha(rs.getString("commit_sha"))
                .build();
    }

    private Map<String, Object> readMetadata(Object rawMetadata) {
        if (rawMetadata == null) {
            return null;
        }
        String json = String.valueOf(rawMetadata);
        if (json == null || json.isBlank()) {
            return null;
        }
        try {
            return objectMapper.readValue(json, new TypeReference<>() {
            });
        } catch (Exception e) {
            throw new IllegalStateException("Failed to parse wiki chunk metadata", e);
        }
    }

    private void setVectorSearchEf(java.sql.Connection connection) throws SQLException {
        try (Statement statement = connection.createStatement()) {
            statement.execute("SET hnsw.ef_search = " + HNSW_EF_SEARCH);
        }
    }

    private void resetVectorSearchEf(java.sql.Connection connection) throws SQLException {
        try (Statement statement = connection.createStatement()) {
            statement.execute("RESET hnsw.ef_search");
        }
    }
}

File: wiki/repository-jpa/src/main/kotlin/kr/devport/api/domain/wiki/repository/WikiSectionChunkRepositoryImpl.kt
package kr.devport.api.domain.wiki.repository

import com.fasterxml.jackson.core.type.TypeReference
import com.fasterxml.jackson.databind.ObjectMapper
import kr.devport.api.domain.wiki.entity.WikiSectionChunk
import kr.devport.api.domain.wiki.repository.WikiSectionChunkRepositoryCustom.ScoredChunkRow
import org.springframework.jdbc.core.ConnectionCallback
import org.springframework.jdbc.core.JdbcTemplate
import org.springframework.jdbc.core.RowMapper
import org.springframework.stereotype.Repository
import java.sql.Connection
import java.sql.ResultSet

@Repository
class WikiSectionChunkRepositoryImpl(
    private val jdbcTemplate: JdbcTemplate,
) : WikiSectionChunkRepositoryCustom {
    private val objectMapper = ObjectMapper()

    private val scoredChunkRowMapper =
        RowMapper { rs, _ -> ScoredChunkRow(mapChunk(rs), rs.getDouble("score")) }

    override fun findSimilarChunksWithScore(
        projectExternalId: String,
        queryEmbedding: String,
        limit: Int,
    ): List<ScoredChunkRow> =
        jdbcTemplate.execute(
            ConnectionCallback { connection ->
                connection.prepareStatement(VECTOR_BY_PROJECT_SQL).use { statement ->
                    statement.setString(1, queryEmbedding)
                    statement.setString(2, projectExternalId)
                    statement.setString(3, queryEmbedding)
                    statement.setInt(4, limit)
                    statement.executeQuery().use { rs -> mapRows(rs) }
                }
            },
        )!!

    override fun findSimilarChunksGlobalWithScore(
        queryEmbedding: String,
        limit: Int,
    ): List<ScoredChunkRow> =
        jdbcTemplate.execute(
            ConnectionCallback { connection ->
                setVectorSearchEf(connection)
                try {
                    connection.prepareStatement(VECTOR_GLOBAL_SQL).use { statement ->
                        statement.setString(1, queryEmbedding)
                        statement.setString(2, queryEmbedding)
                        statement.setInt(3, limit)
                        statement.executeQuery().use { rs -> mapRows(rs) }
                    }
                } finally {
                    resetVectorSearchEf(connection)
                }
            },
        )!!

    override fun findLexicalCandidates(
        projectExternalId: String,
        question: String,
        limit: Int,
    ): List<ScoredChunkRow> =
        jdbcTemplate.query(
            LEXICAL_SQL,
            scoredChunkRowMapper,
            question,
            question,
            projectExternalId,
            question,
            question,
            limit,
        )

    private fun mapRows(rs: ResultSet): List<ScoredChunkRow> {
        val rows = mutableListOf<ScoredChunkRow>()
        var rowNum = 0
        while (rs.next()) {
            rows.add(scoredChunkRowMapper.mapRow(rs, rowNum++)!!)
        }
        return rows
    }

    private fun mapChunk(rs: ResultSet): WikiSectionChunk =
        WikiSectionChunk().apply {
            id = rs.getLong("id")
            projectExternalId = rs.getString("project_external_id")
            sectionId = rs.getString("section_id")
            subsectionId = rs.getString("subsection_id")
            chunkType = rs.getString("chunk_type")
            content = rs.getString("content")
            tokenCount = rs.getInt("token_count")
            metadata = readMetadata(rs.getObject("metadata"))
            commitSha = rs.getString("commit_sha")
        }

    private fun readMetadata(rawMetadata: Any?): MutableMap<String, Any>? {
        if (rawMetadata == null) {
            return null
        }
        val json = rawMetadata.toString()
        if (json.isBlank()) {
            return null
        }
        return try {
            objectMapper.readValue(json, object : TypeReference<MutableMap<String, Any>>() {})
        } catch (e: Exception) {
            throw IllegalStateException("Failed to parse wiki chunk metadata", e)
        }
    }

    private fun setVectorSearchEf(connection: Connection) {
        connection.createStatement().use { it.execute("SET hnsw.ef_search = $HNSW_EF_SEARCH") }
    }

    private fun resetVectorSearchEf(connection: Connection) {
        connection.createStatement().use { it.execute("RESET hnsw.ef_search") }
    }

    companion object {
        private const val HNSW_EF_SEARCH = 64

        private const val SELECT_CHUNK_COLUMNS = """
            SELECT
                c.id,
                c.project_external_id,
                c.section_id,
                c.subsection_id,
                c.chunk_type,
                c.content,
                c.token_count,
                c.metadata,
                c.commit_sha
            """

        private const val VECTOR_BY_PROJECT_SQL =
            SELECT_CHUNK_COLUMNS +
                """
                , 1 - (c.embedding <=> cast(? AS vector)) AS score
                FROM wiki_section_chunks c
                WHERE c.project_external_id = ?
                ORDER BY c.embedding <=> cast(? AS vector)
                LIMIT ?
                """

        private const val VECTOR_GLOBAL_SQL =
            SELECT_CHUNK_COLUMNS +
                """
                , 1 - (c.embedding <=> cast(? AS vector)) AS score
                FROM wiki_section_chunks c
                ORDER BY c.embedding <=> cast(? AS vector)
                LIMIT ?
                """

        private const val LEXICAL_SQL =
            SELECT_CHUNK_COLUMNS +
                """
                ,
                (similarity(c.content, ?) + 0.5 * COALESCE(similarity(c.metadata->>'titleKo', ?), 0)) AS score
                FROM wiki_section_chunks c
                WHERE c.project_external_id = ?
                  AND (c.content % ? OR c.metadata->>'titleKo' % ?)
                ORDER BY score DESC
                LIMIT ?
                """
    }
}
